### ISSUE TITLE
Tainting node using ignored taint causes node groups to become unhealthy

### ISSUE THREAD

**Issue by wcurry** [kubernetes/autoscaler#3985]
**Which component are you using?**:

cluster-autoscaler

**What version of the component are you using?**:

v1.18.3

Component version:

**What k8s version are you using (`kubectl version`)?**:

v1.18.16

**What environment is this in?**:

AWS

**What did you expect to happen?**:

We taint all of our nodes with `recycle=True:NoSchedule`. A node recycling service proceeds to drain these nodes and terminate them one at a time. Sometimes this process slows down due to misconfigured PDBs. During that time pending pods start stacking up due to modified deployments. Scale-up should occur as pending pods pile up and replacement nodes don't show up in time.

**What happened instead?**:

We see the node groups show `unready` and all of the nodes that were `ready` appear in the `longNotStarted` group. I captured all non-master nodes' json while the taint was present and CA is reporting those nodes in the `longNotStarted` group and checked the conditions evaluated in `isNodeStillStarting()` and every node is fine and should not be marked `longNotStarted`.

**How to reproduce it (as minimally and precisely as possible)**:

Our args:

```
            - ./cluster-autoscaler
            - --balance-similar-node-groups
            - --cloud-provider=aws
            - --expander=least-waste
            - --expendable-pods-priority-cutoff=0
            - --ignore-taint=recycle
            - --max-autoprovisioned-node-group-count=15
            - --max-empty-bulk-delete=2
            - --max-failing-time=15m
            - --max-graceful-termination-sec=600
            - --max-inactivity=10m
            - --max-node-provision-time=15m
            - --max-total-unready-percentage=33
            - --node-autoprovisioning-enabled=false
            - --node-group-auto-discovery=asg:tag=k8s.io/cluster-autoscaler/enabled,kubernetes.io/cluster/$cluster_id
            - --ok-total-unready-count=3
            - --skip-nodes-with-local-storage=false
            - --skip-nodes-with-system-pods=false
            - --scale-down-candidates-pool-min-count=50
            - --scale-down-candidates-pool-ratio=0.1
            - --scale-down-delay-after-add=10m
            - --scale-down-delay-after-delete=0m
            - --scale-down-delay-after-failure=3m
            - --scale-down-enabled=true
            - --scale-down-non-empty-candidates-count=30
            - --scale-down-unneeded-time=10m
            - --scale-down-unready-time=20m
            - --scale-down-utilization-threshold=0.50
            - --scan-interval=1m
            - --stderrthreshold=info
            - --v=1
            - --write-status-configmap=true
```

**Anything else we need to know?**:

```
W0401 20:31:31.546160       1 scale_up.go:380] Node group cluster-1-general_nodes-asg-us-west-2a is not ready for scaleup - unhealthy
W0401 20:31:31.546190       1 scale_up.go:380] Node group cluster-1-general_nodes-asg-us-west-2b is not ready for scaleup - unhealthy
W0401 20:31:31.546240       1 scale_up.go:380] Node group cluster-1-general_nodes-asg-us-west-2c is not ready for scaleup - unhealthy
W0401 20:31:31.546259       1 scale_up.go:380] Node group cluster-1-general_pool_b_nodes-asg-us-west-2a is not ready for scaleup - unhealthy
W0401 20:31:31.546279       1 scale_up.go:380] Node group cluster-1-general_pool_b_nodes-asg-us-west-2b is not ready for scaleup - unhealthy
W0401 20:31:31.546297       1 scale_up.go:380] Node group cluster-1-general_pool_b_nodes-asg-us-west-2c is not ready for scaleup - unhealthy
I0401 20:31:31.546317       1 scale_up.go:437] No expansion options
```

```
Name:        cluster-1-general_nodes-asg-us-west-2b
 Health:      Unhealthy (ready=0 unready=0 notStarted=0 longNotStarted=44 registered=44 longUnregistered=0 cloudProviderTarget=44 (minSize=5, maxSize=90))
              LastProbeTime:      2021-04-01 20:31:31.032507223 +0000 UTC m=+2496.916940009
              LastTransitionTime: 2021-04-01 20:29:29.965337725 +0000 UTC m=+2375.849770587
 ScaleUp:     Backoff (ready=0 cloudProviderTarget=44)
              LastProbeTime:      2021-04-01 20:31:31.032507223 +0000 UTC m=+2496.916940009
              LastTransitionTime: 2021-04-01 20:29:29.965337725 +0000 UTC m=+2375.849770587
 ScaleDown:   NoCandidates (candidates=0)
              LastProbeTime:      2021-04-01 20:31:31.032507223 +0000 UTC m=+2496.916940009
              LastTransitionTime: 2021-04-01 19:51:14.406068686 +0000 UTC m=+80.290501545

 Name:        cluster-1-general_nodes-asg-us-west-2c
 Health:      Unhealthy (ready=0 unready=0 notStarted=0 longNotStarted=38 registered=38 longUnregistered=0 cloudProviderTarget=38 (minSize=5, maxSize=90))
              LastProbeTime:      2021-04-01 20:31:31.032507223 +0000 UTC m=+2496.916940009
              LastTransitionTime: 2021-04-01 20:29:29.965337725 +0000 UTC m=+2375.849770587
 ScaleUp:     Backoff (ready=0 cloudProviderTarget=38)
              LastProbeTime:      2021-04-01 20:31:31.032507223 +0000 UTC m=+2496.916940009
              LastTransitionTime: 2021-04-01 20:29:29.965337725 +0000 UTC m=+2375.849770587
 ScaleDown:   NoCandidates (candidates=0)
              LastProbeTime:      2021-04-01 20:31:31.032507223 +0000 UTC m=+2496.916940009
              LastTransitionTime: 2021-04-01 19:51:14.406068686 +0000 UTC m=+80.290501545

 Name:        cluster-1-general_pool_b_nodes-asg-us-west-2a
 Health:      Unhealthy (ready=3 unready=0 notStarted=0 longNotStarted=8 registered=11 longUnregistered=0 cloudProviderTarget=12 (minSize=0, maxSize=20))
              LastProbeTime:      2021-04-01 20:31:31.032507223 +0000 UTC m=+2496.916940009
              LastTransitionTime: 2021-04-01 20:29:29.965337725 +0000 UTC m=+2375.849770587
 ScaleUp:     Backoff (ready=3 cloudProviderTarget=12)
              LastProbeTime:      2021-04-01 20:31:31.032507223 +0000 UTC m=+2496.916940009
              LastTransitionTime: 2021-04-01 20:29:29.965337725 +0000 UTC m=+2375.849770587
 ScaleDown:   NoCandidates (candidates=0)
              LastProbeTime:      2021-04-01 20:31:31.032507223 +0000 UTC m=+2496.916940009
              LastTransitionTime: 2021-04-01 19:51:14.406068686 +0000 UTC m=+80.290501545

 Name:        cluster-1-general_pool_b_nodes-asg-us-west-2b
 Health:      Unhealthy (ready=0 unready=0 notStarted=0 longNotStarted=10 registered=10 longUnregistered=0 cloudProviderTarget=10 (minSize=0, maxSize=20))
              LastProbeTime:      2021-04-01 20:31:31.032507223 +0000 UTC m=+2496.916940009
              LastTransitionTime: 2021-04-01 20:29:29.965337725 +0000 UTC m=+2375.849770587
 ScaleUp:     Backoff (ready=0 cloudProviderTarget=10)
              LastProbeTime:      2021-04-01 20:31:31.032507223 +0000 UTC m=+2496.916940009
              LastTransitionTime: 2021-04-01 20:29:29.965337725 +0000 UTC m=+2375.849770587
 ScaleDown:   NoCandidates (candidates=0)
              LastProbeTime:      2021-04-01 20:31:31.032507223 +0000 UTC m=+2496.916940009
              LastTransitionTime: 2021-04-01 19:51:14.406068686 +0000 UTC m=+80.290501545

 Name:        cluster-1-general_pool_b_nodes-asg-us-west-2c
 Health:      Unhealthy (ready=0 unready=0 notStarted=0 longNotStarted=11 registered=11 longUnregistered=0 cloudProviderTarget=11 (minSize=0, maxSize=20))
              LastProbeTime:      2021-04-01 20:31:31.032507223 +0000 UTC m=+2496.916940009
              LastTransitionTime: 2021-04-01 20:29:29.965337725 +0000 UTC m=+2375.849770587
 ScaleUp:     Backoff (ready=0 cloudProviderTarget=11)
              LastProbeTime:      2021-04-01 20:31:31.032507223 +0000 UTC m=+2496.916940009
              LastTransitionTime: 2021-04-01 20:29:29.965337725 +0000 UTC m=+2375.849770587
 ScaleDown:   NoCandidates (candidates=0)
              LastProbeTime:      2021-04-01 20:31:31.032507223 +0000 UTC m=+2496.916940009
              LastTransitionTime: 2021-04-01 19:51:14.406068686 +0000 UTC m=+80.290501545
```

**Comment by wcurry** [comment]
Tainting just one node earlier in the day didn't reproduce the unhealthy state. We may try progressively larger sets of tainted nodes to see where it breaks down.

**Comment by MaciekPytel** [comment]
This is actually expected, unfortunately `--ignore-taint` isn't really doing what the name implies. The _only_ use-case supported by ignore-taint is one where a node requires additional custom initialization (ex. installing drivers, starting some DS) before it can accept pods. In this case node can be started with ignore-taint and the taint can be removed once the initialization is done. 

As long as the node has an ignore-taint CA will treat it as still booting up / unready. This is needed to avoid infinite scale-up (node is created with ignore-taint, pods remain pending, CA immediately triggers scale-up again, this cycle repeats until the taint is removed from some nodes). Since a node with ignore-taint is unready as far as CA is concerned, once there are enough such nodes in the NodeGroup CA will start treating the NG as unhealthy (thresholds are controlled by `--max-total-unready-percentage`, `--ok-total-unready-count`). 

It may or may not be ok to use ignore-taints for some other use-cases that involve tainting nodes for a very short time, but it's not really supported and it will very much depend on the details of use-case.

---

The name of the flag is clearly very misleading, this is because of historical reasons: the initial desire was for CA to just ignore the existence of the taint. Unfortunately it turned out CA logic depends much more heavily on nodes in NG being identical for purposes of scheduling and just filtering out the taint from template node leads to CA completely misunderstanding the state of the cluster and issues like infinite scale-up described above. The flag remained as it works well for the use-case of custom node initialization (and it's a popular use-case).

**Comment by wcurry** [comment]
Thank you for the clarification. Are you aware of any workarounds for this pattern of node recycling? Immediately I was thinking of setting the `--ok-total-unready-count` to somewhere around 30 and batching the recycling taint on the oldest ~24 nodes (less to avoid zone imbalances marking a group offline).

**Comment by matthias50** [comment]
@MaciekPytel, I know this issue is closed but the above explanation that `--ignore-taints` doesn't match with what was originally claimed in the PR and I'd like clarification. In the PR https://github.com/kubernetes/autoscaler/pull/1963, it mentions that it fixes #1350.  Issue #1350 is precisely this use case of node recycling and we have been relying on being able to use --ignore-taints for the node recycling case since #1350 was closed. To me, it seems that this issue was closed in error since, from my perspective, `--ignore-taints` is no longer addressing #1350.
